flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.data.binary.BinaryStringData;
+import org.apache.flink.table.data.binary.BinaryStringDataUtil;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.util.Iterator;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_JOIN}. */
+@Internal
+public class ArrayJoinFunction extends BuiltInScalarFunction {
+
+    private final ArrayData.ElementGetter elementGetter;
+
+    public ArrayJoinFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_JOIN, context);
+        final DataType elementDataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType();
+        elementGetter = ArrayData.createElementGetter(elementDataType.getLogicalType());
+    }
+
+    public @Nullable StringData eval(
+            ArrayData array, StringData delimiter, StringData... nullReplacement) {
+        try {
+            if (array == null
+                    || delimiter == null
+                    || (nullReplacement.length != 0 && nullReplacement[0] == null)) {
+                return null;
+            }
+            return BinaryStringDataUtil.concatWs(
+                    (BinaryStringData) delimiter,
+                    () -> new ArrayIterator(elementGetter, array, nullReplacement));
+        } catch (Throwable t) {
+            throw new FlinkRuntimeException(t);
+        }
+    }
+
+    static final class ArrayIterator implements Iterator<BinaryStringData> {
+
+        private final int size;
+        private final ArrayData.ElementGetter elementGetter;
+        private final ArrayData arrayData;
+        private final StringData[] nullReplacement;
+        private int currentPos;
+
+        public ArrayIterator(
+                ArrayData.ElementGetter elementGetter,
+                ArrayData arrayData,
+                StringData[] nullReplacement) {
+            this.size = arrayData.size();
+            this.elementGetter = elementGetter;
+            this.arrayData = arrayData;
+            this.nullReplacement = nullReplacement;
```

Comment:
why don't we pass just the first element of an array instead? In the current approach we have unnecessary array indexing.

Reply:
Still you should perform all those checks before creating the iterator. It does not change between calls to `next`.

You can do sth like:
```
if (nullReplacement.length != 0 && nullReplacement[0] != null) {
 normalizedReplacement = nullReplacement[0]  
} else {
  normalizedReplacement = null
}

new ArrayIterator(elementGetter, array, normalizedReplacement)
```